ColorPicker Updates

robloo opened the pull request:
## What does the pull request do?

Updates the ColorPicker group of controls with more fixes/improvements.

 1. A special case was added to handle #00000000 in the ColorSpectrum upon first selection. If this case is detected the third component and alpha component will automatically be set to maximum values. Full reasoning is commented in code.
 1. ColorSlider.`IsAlphaMaxForced` was renamed to `IsAlphaVisible`. This follows terminology in other controls/primitives and essentially does the same as before.
 1. One of the largest changes is ColorSlider.`IsSaturationValueMaxForced` being replaced with the more powerful/general-purpose `IsPerceptive` property with enhanced functionality. Before this property only worked in Hsva color model, now it also works in Rgba. This allows switching the spectrum slider gradient to render as mathematically correct vs. better for user perception.
      1. By default this changes the look of the RGB color model sliders:
          ![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/cfb72fe7-15f2-4537-a44f-5dcc8d0c0be0)
          They now no longer change -- the gradients will always be red/green/blue in hue.
      1. Alpha slider stays the same and HSV sliders stay the same.
      1.  When this property is set should be changed in the styles/control themes. However, it doesn't appear to work #11271. The API was updated now so this can be addressed later though as a non-breaking change.
 1. Added direct conversions for HsvColor.ToHsl() and HslColor.ToHsv()
 1. Other minor improvements to comments

It was planned to adjust the ColorPreviewer accent color calculation to match the Fluent theme (HSL color model algorithm). However, there were two issues found that stopped this:
 * The algorithm is not reversible, reducing by one step and then increasing by one step will result in a different color than the start.
 * For some reason, hue information is lost when the HSL color reaches minimum/maximum lightness. In HSL this becomes black and white. This then switches the base color to pure black and white with no hue information. I'm not actually sure what is causing this as I added in full conversions between HSV/HSL without going through RGB so information shouldn't be lost. Its just something about HSL I need to think about further I guess.

cc @maxkatz6 

## How was the solution implemented (if it's not obvious)?

As described.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- ~[ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation~

## Breaking changes

Yes, breaking ColorSlider properties from previous previews. These are so far unreleased in stable versions of Avalonia and are only used in the primitive ColorSlider control.
 1. ColorSlider.`IsAlphaMaxForced` renamed to `IsAlphaVisible`
 4. ColorSlider.`IsSaturationValueMaxForced` replaced with more powerful/general-purpose `IsPerceptive` property with enhanced functionality

## Obsoletions / Deprecations

None

## Fixed issues

Fixes #11059

robloo committed:
Add special coercion logic to initial ColorSpectrum selection

This specially handles #00000000 to be much more intuitive for end-users.

robloo committed:
Fix Hue in new ColorSpectrum coercion logic

robloo committed:
Always force max saturation and value in the third component slider

robloo committed:
Rename ColorSlider.IsAlphaMaxForced to IsAlphaVisible

robloo committed:
Replace ColorSlider.IsSaturationValueMaxForced with more powerful/general-purpose IsPerceptive

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0034358-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Improve ColorPreviewer margins with drop shadow

This sets ClipToBounds to false for the control which allows the drop shadow to fully draw above/below the control itself. Doing this removes the need to adjust margins in almost all places. This is a big simplification to margin calculation and removes strange negative values.

robloo committed:
Improve comments

robloo committed:
Merge branch 'master' into colorpicker-updates-7

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorPicker.xaml:
```diff
@@ -490,18 +490,33 @@
       </ControlTemplate>
     </Setter>
+
+    <Style Selector="^ /template/ primitives|ColorSlider#ColorSpectrumThirdComponentSlider[ColorComponent=Component1]">
+      <Setter Property="IsPerceptive" Value="True" />
+    </Style>
+
+    <Style Selector="^ /template/ primitives|ColorSlider#Component1Slider[ColorModel=Rgba]">
+      <Setter Property="IsPerceptive" Value="False" />
+    </Style>
+    <Style Selector="^ /template/ primitives|ColorSlider#Component2Slider[ColorModel=Rgba]">
+      <Setter Property="IsPerceptive" Value="False" />
+    </Style>
+    <Style Selector="^ /template/ primitives|ColorSlider#Component3Slider[ColorModel=Rgba]">
+      <Setter Property="IsPerceptive" Value="False" />
+    </Style>
```
There seems to be an issue here: https://github.com/AvaloniaUI/Avalonia/issues/11271

robloo replied:
Removed for now.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0034364-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo commented on src/Avalonia.Controls.ColorPicker/ColorSlider/ColorSlider.Properties.cs:
```diff
@@ -109,14 +109,39 @@ public HsvColor HsvColor
+        /// Also note this property will only adjust components other than <see cref="ColorComponent"/> during rendering.
+        /// This doesn't change the values of any components in the color – it is only for display.
+        /// </remarks>
+        public bool IsPerceptive
```
The term "perceptive" is used here as it indicates the control should improve "user-perception". "perceptive" perhaps doesn't make the most sense in the context of the control itself though. The ColorSlider isn't "perceptive".

If anyone has concerns with this property naming please let me know and if you have a better idea. I can't really find examples of this concept anywhere.

robloo committed:
Merge branch 'master' into colorpicker-updates-7

robloo committed:
Always enable ColorSlider.IsPerceptive

This works-around #11271 and stabilizes the API. It can be revisited later.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0034606-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo committed:
Remove static color conversion methods

This functionality is already provided by, for example, HsvColor.ToRgb(). Having these separate methods just clutters the API.

robloo committed:
Add direct conversions for HsvColor.ToHsl() and HslColor.ToHsv()

robloo committed:
Switch ColorPreviewer's AccentColorConverter to use the FluentTheme's HSL algorithm

robloo committed:
Revert "Switch ColorPreviewer's AccentColorConverter to use the FluentTheme's HSL algorithm"

This reverts commit a91571aa49c025b96a4a280c64d1f6e9ba8fdea1.

robloo committed:
Improve IsPerceptive comments

robloo committed:
Merge branch 'master' into colorpicker-updates-7

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0034672-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

robloo commented:
@maxkatz6 This one should be ready for review.

robloo committed:
Merge branch 'master' into colorpicker-updates-7

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0034681-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
